### ISSUE TITLE
Expat CMake configure error

### ISSUE THREAD

**Issue by lowy** [AcademySoftwareFoundation/OpenColorIO#1415]
debian unstable 
-- GLVND supported
-- Setting namespace to 'OpenColorIO_v2_1dev' as none was specified.
CMake Deprecation Warning at CMakeLists.txt:216 (message):
  Option OCIO_BUILD_SHARED is deprecated.  Please use the cmake standard
  BUILD_SHARED_LIBS=ON (default ON)

-- Could NOT find pystring (missing: pystring_INCLUDE_DIR pystring_LIBRARY) (Required is at least version "1.1.3")
-- Installing pystring: /home/lowy/OpenColorIO/build/ext/dist/lib/libpystring.a (version "1.1.3")
-- OpenImageIO includes     = /usr/include
-- OpenImageIO libraries    = /usr/lib/x86_64-linux-gnu/libOpenImageIO.so
-- OpenImageIO library_dirs = /usr/lib/x86_64-linux-gnu
-- OpenImageIO oiiotool     = OIIOTOOL_BIN-NOTFOUND
CMake Warning at src/bindings/python/CMakeLists.txt:44 (message):
  Building PyOpenColorIO with OCIO_BUILD_DOCS disabled will result in
  incomplete Python docstrings.

-- Configuring done
CMake Error in tests/cpu/CMakeLists.txt:
  Imported target "expat::expat" includes non-existent path

    "/usr/lib/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

CMake Error in tests/cpu/CMakeLists.txt:
  Imported target "expat::expat" includes non-existent path

    "/usr/lib/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

CMake Error in src/OpenColorIO/CMakeLists.txt:
  Imported target "expat::expat" includes non-existent path

    "/usr/lib/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

CMake Error in src/OpenColorIO/CMakeLists.txt:
  Imported target "expat::expat" includes non-existent path

    "/usr/lib/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.

**Comment by Begasus** [comment]
> Imported target "expat::expat" includes non-existent path

Same error here on Haiku, had to comment out the line in `src/OpenColorIO/CMakeLists.txt`

**Comment by hodoulp** [comment]
Hi @Begasus , I cannot test in your context (only access to `macOS`, `ubuntu` (docker) and `centOS` (docker)). But I was expecting that some one on `debian` could help you, but no response. Did you make some progress?

**Comment by Simran-B** [comment]
I reproduced the problem under Ubuntu: `docker run -it ubuntu:21.10`

```
apt update
apt install -y build-essential git cmake libexpat1-dev
cd /home
git clone --depth 1 https://github.com/AcademySoftwareFoundation/OpenColorIO
cd OpenColorIO
mkdir build
cd build
cmake .. -DOCIO_BUILD_APPS=OFF -DOCIO_BUILD_GPU_TESTS=OFF -DOCIO_BUILD_PYTHON=OFF
```

The problem is in `/usr/lib/x86_64-linux-gnu/cmake/expat-2.4.1/expat.cmake`:

```cmake
# Compute the installation prefix relative to this file.
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
if(_IMPORT_PREFIX STREQUAL "/")
  set(_IMPORT_PREFIX "")
endif()

# Create imported target expat::expat
add_library(expat::expat SHARED IMPORTED)

set_target_properties(expat::expat PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)
```

The script starts at the current folder, then goes 3 folders up to determine the import prefix  `/usr/lib`. This is correct for the library that is located at `/usr/lib/x86_64-linux-gnu/libexpatw.so.1.8.1`. However, the same prefix is used to append `/include` which results in `/usr/lib/include` but that folder does not even exist. The main header is at `/usr/include/expat.h`.

Changing the .cmake file to go one folder up further to construct the include path fixes the problem:

```diff
+get_filename_component(_IMPORT_PREFIX_PARENT "${_IMPORT_PREFIX}" PATH) # headers in /usr/include, not /usr/lib/include
 set_target_properties(expat::expat PROPERTIES
-  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
+  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX_PARENT}/include"
 )
```

Note that if you re-use the `_IMPORT_PREFIX` variable instead, then the library won't be found anymore (not in `/usr/x86_64-linux-gnu`) because that variable appears to be used by a script that gets included (not immediately apparent).

I don't know whether this is an expat bug or if the layout of the installed files is somehow non-standard (not a Linux person) and should be fixed by the maintainer who packages this library.

**Comment by elsandosgrande** [comment]
I filed a bug report over on Launchpad for the `expat` source package: https://bugs.launchpad.net/ubuntu/+source/expat/+bug/1956202

**Comment by Begasus** [comment]
Hi @hodoulp someone in our team changed the value for expat in a later commit then mine which seems to fix the issue so far (https://github.com/haikuports/haikuports/commit/4e1e2894c19bfb8de6e6496cf64ab0429d126afd#diff-275d3ac5bcfddc1bc6fe57977c22b2ad0c495e39243e032827ba4ce1853796ff).
I also made adjustments for the installation of OCIO to follow a bit more the gnuinstalldirs. The recipe and patches can be found here: https://github.com/haikuports/haikuports/pull/6086/files
Need to have a look at your patch @Simran-B if it fixes the same here.

**Comment by remia** [comment]
Doing some digging I found issues for the same error:
* https://github.com/libexpat/libexpat/issues/501
* https://github.com/NixOS/nixpkgs/pull/128997

My understanding is that ubuntu use autotools to configure / build expat and pass a custom lib folder `--libdir=\$${prefix}/lib/$(DEB_HOST_MULTIARCH)` which is not correctly reflected in the CMake config script generated later. @hartwork is already aware of the issue and said the easiest fix is on the packaging side.

**Comment by hartwork** [comment]
@remia I confirm. If someone has resources to team up with me on a fix upstream, please get in touch.

**Comment by doug-walker** [comment]
Closing as I believe this was addressed by PR #1578.
